R/teal_data_module-eval_code.R

```diff
@@ -0,0 +1,87 @@
+setOldClass("teal_data_module")
+
+#' @name eval_code
+#' @description
+#' Given code is evaluated in the `qenv` environment of `teal_data` reactive defined in `teal_data_module`.
+#'
+#' @param object (`teal_data_module`)
+#' @inherit teal.code::eval_code
+#' @importMethodsFrom teal.code eval_code
+#' @export
+#' @examples
+#' tdm <- teal_data_module(
+#'   ui = function(id) div(id = shiny::NS(id)("div_id")),
+#'   server = function(id) {
+#'     shiny::moduleServer(id, function(input, output, session) {
+#'       shiny::reactive(teal_data(IRIS = iris))
+#'     })
+#'   }
+#' )
+#' \dontrun{
+#' eval_code(tdm, "IRIS <- subset(IRIS, Species == 'virginica')")
+#' }
+setMethod("eval_code", signature = c("teal_data_module", "character"), function(object, code) {
+  teal_data_module(
+    ui = function(id) {
+      ns <- NS(id)
+      object$ui(ns("mutate_inner"))
+    },
+    server = function(id) {
+      moduleServer(id, function(input, output, session) {
+        data <- object$server("mutate_inner")
+        reactive(eval_code(data(), code))
+      })
+    }
+  )
+})
+
+#' @rdname eval_code
+#' @export
+setMethod("eval_code", signature = c("teal_data_module", "language"), function(object, code) {
+  eval_code(object, code = format_expression(code))
+})
+
+#' @rdname eval_code
+#' @export
+setMethod("eval_code", signature = c("teal_data_module", "expression"), function(object, code) {
+  eval_code(object, code = format_expression(code))
+})
+
+#' @inherit teal.code::within.qenv params title details
+#' @description
+#' Convenience function for evaluating inline code inside the environment of a
+#' `teal_data_module`
+#'
+#' @param data (`teal_data_module`) object
+#' @return Returns a `teal_data_module` object with a delayed evaluation of `expr`
+#' when module.
+#' @export
+#' @seealso [base::within()], [teal_data_module()]
+#' @examples
+#' tdm <- teal_data_module(
+#'   ui = function(id) div(id = shiny::NS(id)("div_id")),
+#'   server = function(id) {
+#'     shiny::moduleServer(id, function(input, output, session) {
+#'       shiny::reactive(teal_data(IRIS = iris))
+#'     })
+#'   }
+#' )
+#' \dontrun{
+#' within(tdm, IRIS <- subset(IRIS, Species == "virginica"))
+#' }
+within.teal_data_module <- function(data, expr, ...) {
+  expr <- substitute(expr)
+  extras <- list(...)
+
+  # Add braces for consistency.
+  if (!identical(as.list(expr)[[1L]], as.symbol("{"))) {
+    expr <- call("{", expr)
+  }
+
+  calls <- as.list(expr)[-1]
+
+  # Inject extra values into expressions.
+  calls <- lapply(calls, function(x) do.call(substitute, list(x, env = extras)))
+
+  eval_code(object = data, code = as.expression(calls))
+}
```

Comment:
The body is a copy from `teal.code:::within.qenv`, so in practice it could be replaced with code below _(reducing duplicated code accross `{teal.*}`)_:

Thoughts?

```suggestion
within.teal_data_module <- function(data, expr, ...) {
  within.qenv <- getFromNamespace("within.qenv", "teal.code")
  within.qenv(data, expr, ...)
}
```

Reply:
1. I think `getS3method("within", "qenv", envir = as.environment(getNamespace("teal.code")))` would be a better choice.
2. It shouldn't be necessary anyway because `teal.code` must be installed when `teal` is.
3. At the moment `within` only processes the expression and calls `eval_code`. In this context this is a valid simplification. I'm not sure, though :thinking:

Reply:
> 1. I think getS3method("within", "qenv", envir = as.environment(getNamespace("teal.code"))) would be a better choice.
> 2. It shouldn't be necessary anyway because teal.code must be installed when teal is.

That's better! Is the `envir` parameter required? _(as you said, `teal.code` is imported and it seems to find the within)_

> 3. At the moment within only processes the expression and calls eval_code. In this context this is a valid simplification. I'm not sure, though 🤔

@vedhav brough a smilar concern, I'm on the fence as I prefer to minimize repetitive code, but I don't think this solution is "best practice" (calling `within.qenv` with a different data object seems like a red flag). 

I would prefer if there was `within.default` (non-exported) or `.generic_within` in `{teal.code}` that was the backbone of both of those calls and still allow for deviations in the future of `within.qenv` that wouldn't affect `within.teal_data_module`.

Reply:
> That's better! Is the `envir` parameter required?

It isn't but just in case there is a _different_ `within.qenv` defined somewhere...

Reply:
> I would prefer if there was `within.default`

The `within` generic is a `base` function so defining a `default` that suits our particular classes would cause a mess.

Reply:
Using the `.default` was a suggestion if it didn't interfere with `base::within` (by not exporting it, but even then it might be "dangerous")

Reply:
Ah, but S3 methods must be exported :wink:

Reply:
One interesting pattern that I have found when working on the `verdepcheck` is to make a _simple_ (i.e. non S3 / non S4 etc.) (implementation) function for each method and these methods are essentially a one-liner calling those functions. Please find an example in R6 context: https://github.com/r-lib/pkgdepends/blob/main/R/pkg-plan.R - all public methods are calling private (implementation) functions.

Then you can re-use those implementation functions from multiple classes but there are some obvious drawbacks though for package size, documentation, maintenance, debugging etc.

This however is not addressing an exporting / importing issue mentioned here 😈